contracts/metatx/MinimalForwarder.sol

```diff
@@ -26,29 +21,64 @@ contract MinimalForwarder is EIP712 {
         bytes data;
     }
 
-    bytes32 private constant _TYPEHASH =
+    struct ExpiringForwardRequest {
+        uint48 deadline;
+        address from;
+        address to;
+        uint256 value;
+        uint256 gas;
+        uint256 nonce;
+        bytes data;
+    }
+
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
         keccak256("ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,bytes data)");
+    bytes32 private constant _EXPIRING_FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ExpiringForwardRequest(uint48 deadline,address from,address to,uint256 value,uint256 gas,uint256 nonce,bytes data)"
+        );
 
     mapping(address => uint256) private _nonces;
 
-    constructor() EIP712("MinimalForwarder", "0.0.1") {}
+    constructor() EIP712(name(), version()) {}
+
+    function version() public view virtual returns (string memory) {
+        return "0.0.1";
+    }
 
-    function getNonce(address from) public view returns (uint256) {
+    function name() public view virtual returns (string memory) {
+        return "MinimalForwarder";
+    }
+
+    function getNonce(address from) public view virtual returns (uint256) {
         return _nonces[from];
     }
 
-    function verify(ForwardRequest calldata req, bytes calldata signature) public view returns (bool) {
-        address signer = _hashTypedDataV4(
-            keccak256(abi.encode(_TYPEHASH, req.from, req.to, req.value, req.gas, req.nonce, keccak256(req.data)))
-        ).recover(signature);
-        return _nonces[req.from] == req.nonce && signer == req.from;
+    function verify(ForwardRequest calldata req, bytes calldata signature) public view virtual returns (bool) {
```

Comment:
That is not what we agreed. If we do that old signature are no longer valid. We need to continue verifying `ForwardRequest` using the `ForwardRequest` type.

Reply:
I was exploring how to arrange them together since we need to keep backward compatibility. I'll solve this thread since I feel the focus is on the incorrect verification I pushed as a draft last night and that's solved.

Please open another thread if you think there's something to discuss related to the interface.

Reply:
What I am saying is, by forcing the verification of `ForwardRequest` to be done using the verification of `ExpiringForwardRequest` (which include a different hashing mechanism) you are actually breaking backward compatibility. Not keeping it.

Reply:
It is **not** ok to verify everything using the `ExpiringForwardRequest` typed struct. We need to keep the verfication of "old" `ForwardRequest` using the existing signature scheme.